crypto/evp/m_blake2b.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * BLAKE2 reference source code package - reference C implementations
+ *
+ * Copyright 2012, Samuel Neves <sneves@dei.uc.pt>.
+ * You may use this under the terms of the CC0, the OpenSSL Licence, or the
+ * Apache Public License 2.0, at your option.  The terms of these licenses can
+ * be found at:
+ *
+ * - OpenSSL license   : https://www.openssl.org/source/license.html
+ * - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
+ * - CC0 1.0 Universal : http://www.apache.org/licenses/LICENSE-2.0
```

Comment:
On Wed, Mar 9, 2016 at 3:50 PM, ekasper <notifications@github.com> wrote:

> In crypto/evp/m_blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55611735>:
>
> > @@ -0,0 +1,63 @@
> > +/*
> > + * BLAKE2 reference source code package - reference C implementations
> > + *
> > + * Copyright 2012, Samuel Neves <sneves@dei.uc.pt>.
> > + * You may use this under the terms of the CC0, the OpenSSL Licence, or the
> > + * Apache Public License 2.0, at your option.  The terms of these licenses can
> > + * be found at:
> > + *
> > + * - OpenSSL license   : https://www.openssl.org/source/license.html
> > + * - Apache 2.0        : http://www.apache.org/licenses/LICENSE-2.0
> > + * - CC0 1.0 Universal : http://www.apache.org/licenses/LICENSE-2.0
>
> And to clarify (I explained to Kurt internally as well), the top license
> header MUST be our standard header. But it's fine to add a custom note
> afterwards, for example
>
> /*
>  * Derived from the BLAKE2 reference implementation written by Samuel Neves.
>  * More information about the BLAKE2 hash function and its implementations
>  * can be found at https://blake2.net.
>  */
>
> I can put the standard OpenSSL copyright header on top, with this comment
> below.  It says:
>
I'm happy to add this at the top:

/* ====================================================================
 * Copyright (c) 2015 The OpenSSL Project. All rights reserved.
 *
 * Rights for redistribution and usage in source and binary
 * forms are granted according to the OpenSSL license.
 */

However, is this technically correct, since Samuel Neves wrote it?  He is
willing to give away the code under any conditions we like, so long as it
does not restrict others in any way to also use it.  I don't think he is
giving the code to the OpenSSL project, just granting a license to the
world to use it under 3 different licenses.